File: rest/src/test/java/com/jgoticks/TicketSellerTest.java
package com.jgoticks;

import akka.actor.ActorRef;
import akka.actor.ActorSystem;
import akka.japi.Option;
import akka.pattern.Patterns;
import akka.testkit.javadsl.TestKit;
import org.junit.AfterClass;
import org.junit.Test;

import java.util.Collections;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

import static com.jgoticks.TicketSeller.*;

public class TicketSellerTest extends TestKit {
    private static ActorSystem system_ = ActorSystem.create();

    public TicketSellerTest() {
        super(system_);
    }

    @AfterClass
    public static void teardownClass() throws Exception {
        system_.terminate();
    }

    private List<Ticket> makeTickets(int count) {
        return IntStream.range(0, count).mapToObj(Ticket::new).collect(Collectors.toList());
    }

    @Test
    public void test_add() {
        ActorRef testActor = getSystem().actorOf(TicketSeller.props("event test_add"), "test_add");

        testActor.tell(new Add(makeTickets(3)), getRef());
        expectNoMsg();
    }

    @Test
    public void test_buy() {
        ActorRef testActor = getSystem().actorOf(TicketSeller.props("event test_buy"), "test_buy");
        Patterns.ask(testActor, new Add(makeTickets(3)), 3000);

        testActor.tell(new Buy(1), getRef());
        expectMsg(Tickets.of("event test_buy", makeTickets(1)));

        testActor.tell(new Buy(3), getRef());
        expectMsg(Tickets.of("event test_buy", Collections.emptyList()));
    }

    @Test
    public void test_getEvent() {
        ActorRef testActor = getSystem().actorOf(TicketSeller.props("event test_getEvent"), "test_getEvent");
        Patterns.ask(testActor, new Add(makeTickets(5)), 3000);

        testActor.tell(new GetEvent(), getRef());
        expectMsg(Option.some(new BoxOffice.Event("event test_getEvent", 5)));
    }

    @Test
    public void test_cancel() {
        ActorRef testActor = getSystem().actorOf(TicketSeller.props("event test_cancel"), "test_cancel");
        testActor.tell(new Add(makeTickets(1000)),  getRef());

        testActor.tell(new Cancel(), getRef());
        expectMsg(Option.some(new BoxOffice.Event("event test_cancel",1000)));

        testActor.tell(new Buy(1000),  getRef());
        expectNoMsg();
    }
}

File: rest-remote/src/test/scala/com/goticks/TicketSellerTest.scala
package com.goticks

import akka.actor.ActorSystem
import akka.testkit.{ImplicitSender, TestKit}
import akka.util.Timeout
import com.goticks.BoxOffice.Event
import com.goticks.TicketSeller._
import org.scalatest.{MustMatchers, WordSpecLike}

class TicketSellerTest extends TestKit(ActorSystem("testssytem"))
  with WordSpecLike
  with MustMatchers
  with ImplicitSender
  with StopSystemAfterAll {

  def makeTickets = (1 to 10).map(Ticket.apply).toVector

  "The TicketSeller" must {
    "add ticket when a Add(Ticket(1)) is sent to it" in {
      val seller = system.actorOf(TicketSeller.props("test add event"), "seller_actor_Add")

      seller ! Add(Vector(Ticket(1)))
      expectNoMsg()
    }
    "buy ticket when a Buy(2) if the actor has two tickets, is sent to it" in {
      import akka.pattern.ask

      import scala.concurrent.duration._

      val seller = system.actorOf(TicketSeller.props("test buy event"), "seller_actor_Buy")
      implicit val timeout = Timeout(3 seconds)
      seller.ask(Add(makeTickets))

      seller ! Buy(2)
      expectMsg(Tickets("test buy event", Vector(Ticket(1), Ticket(2))))
    }
    "buy ticket failed when a Buy(2) if the actor has not two tickets, is sent to it" in {
      import akka.pattern.ask
      import scala.concurrent.duration._

      val props = TicketSeller.props("test buy event failed")
      val seller = system.actorOf(props, "seller_actor_Buy_FailCase")
      implicit val timeout = Timeout(3 seconds)
      seller.ask(Add(Vector(Ticket(1))))

      seller ! Buy(2)
      expectMsg(Tickets("test buy event failed", Vector.empty[Ticket]))
    }
    "get event when a GetEvent() is sent to it" in {
      val seller = system.actorOf(TicketSeller.props("test getEvent"), "seller_actor_GetEvent")

      seller ! GetEvent
      expectMsg(Some(Event("test getEvent", 0)))
    }
    "cancel event when a Cancel() is sent to it" in {
      import akka.pattern.ask
      import scala.concurrent.duration._

      val seller = system.actorOf(TicketSeller.props("test cancel"), "seller_actor_Cancel")
      implicit val timeout = Timeout(3 seconds)
      seller.ask(Add(makeTickets))

      seller ! Cancel
      expectMsg(Some(Event("test cancel", 10)))

      seller ! Buy(10)
      expectNoMsg()
    }
  }
}
